src/core/utils/validate-ipni-advertisement.ts

```diff
@@ -126,3 +251,262 @@ export async function validateIPNIAdvertisement(
     })
   })
 }
+
+/**
+ * Convert a PDP service URL to an IPNI multiaddr format.
+ *
+ * Storage providers expose their PDP (Piece Data Provider) service via HTTP/HTTPS
+ * endpoints (e.g., "https://provider.example.com:8443"). When they advertise content
+ * to IPNI, they include multiaddrs in libp2p format (e.g., "/dns/provider.example.com/tcp/8443/https").
+ *
+ * This function converts between these representations to enable validation that a
+ * provider's IPNI advertisement matches their registered service endpoint.
+ *
+ * @param serviceURL - HTTP/HTTPS URL of the provider's PDP service
+ * @param logger - Optional logger for warnings
+ * @returns Multiaddr string in libp2p format, or undefined if conversion fails
+ *
+ * @example
+ * serviceURLToMultiaddr('https://provider.example.com')
+ * // Returns: '/dns/provider.example.com/tcp/443/https'
+ *
+ * @example
+ * serviceURLToMultiaddr('http://provider.example.com:8080')
+ * // Returns: '/dns/provider.example.com/tcp/8080/http'
+ */
+export function serviceURLToMultiaddr(serviceURL: string, logger?: Logger): string | undefined {
+  try {
+    const url = new URL(serviceURL)
+    const port =
+      url.port !== ''
+        ? Number.parseInt(url.port, 10)
+        : url.protocol === 'https:'
+          ? 443
+          : url.protocol === 'http:'
+            ? 80
+            : undefined
+
+    if (Number.isNaN(port) || port == null) {
+      return undefined
+    }
+
+    const protocolComponent =
+      url.protocol === 'https:' ? 'https' : url.protocol === 'http:' ? 'http' : url.protocol.replace(':', '')
+
+    return `/dns/${url.hostname}/tcp/${port}/${protocolComponent}`
+  } catch (error) {
+    logger?.warn({ serviceURL, error }, 'Unable to derive IPNI multiaddr from serviceURL')
+    return undefined
+  }
+}
+
+/**
+ * Extract all provider results from the IPNI gateway response.
+ *
+ * The response can contain multiple multihash results, each with multiple provider
+ * results. This flattens them into a single array for easier processing.
```

Comment:
```suggestion
 * The response can contain multiple providers results, each with multiple multiaddr esults. 
 * This flattens them into a single array for easier processing.
```

I think this is right given what I see in https://cid.contact/cid/bafybeigvgzoolc3drupxhlevdp2ugqcrbcsqfmcek2zxiw5wctk3xjpjwy

```
  "MultihashResults": [
    {
      "Multihash": "EiDVNlzli2ONH3OslRv1Q0BRCKUCsERWs3RbthTVu6Xptg==",
      "ProviderResults": [
        {
          "ContextID": "ghsAAAAIAAAAANgqWCgAAYHiA5IgIPiHIx1wjhx1H/MDYlbZ74WEz8C+Rap56+xsji2aePEe",
          "Metadata": "oBIA",
          "Provider": {
            "ID": "12D3KooWBbMcX6JiT7rcu9gdc9hNLiq4LuVk4a8he8eE2fJBCJAm",
            "Addrs": [
              "/dns/a-fil-http.aur.lu/tcp/443/https"
            ]
          }
        },
        {
          "ContextID": "ghsAAAAIAAAAANgqWCgAAYHiA5IgINWxbqG9f5J52zej06J43+sFBngf6nNoSW+KK4jsEY0z",
          "Metadata": "oBIA",
          "Provider": {
            "ID": "12D3KooWP9jPCTRwpRk7J7CknwqC3h14ujuvtBoN518YuiJ5MSFR",
            "Addrs": [
              "/dns/a-fil-http.aur.lu/tcp/443/https"
            ]
          }
        },
```